llrt_core/src/modules/encoding/text_decoder.rs

```diff
@@ -64,24 +63,31 @@ impl<'js> TextDecoder {
     }
 
     pub fn decode(&self, ctx: Ctx<'js>, buffer: Value<'js>) -> Result<String> {
-        let bytes = get_bytes(&ctx, buffer)?;
-
-        let decoder = Encoding::for_label(self.encoding.as_bytes()).unwrap();
+        let bytes = if buffer.is_undefined() {
+            vec![]
+        } else {
+            get_bytes(&ctx, buffer)?
+        };
 
-        let str: Cow<str>;
-        let has_error: bool;
+        let decoder = Encoding::for_label(self.encoding.as_bytes()).or_throw_msg(
+            &ctx,
+            &["Failed to get decoder for encoding ", &self.encoding].concat(),
+        )?;
 
-        if decoder == encoding_rs::UTF_8 {
-            (str, has_error) = match self.ignore_bom {
-                false => decoder.decode_with_bom_removal(&bytes),
-                true => decoder.decode_without_bom_handling(&bytes),
-            }
-        } else {
-            (str, _, has_error) = decoder.decode(&bytes);
-        }
+        let (str, has_error) = match self.ignore_bom {
+            false => decoder.decode_with_bom_removal(&bytes),
+            true => decoder.decode_without_bom_handling(&bytes),
+        };
```

Comment:
A bit strange to match on bools (I know this was in previous version). I prefer a if:
```rust
let (str, has_error) = if self.ignore_bom {
    decoder.decode_without_bom_handling(&bytes)
} else {
    decoder.decode_with_bom_removal(&bytes)
};
```

Reply:
I could go either way, but I wrote a fair amount of Erlang recently where everything is a `match` so I'm kinda used to seeing it. 🙃